generator/internal/sidekick/command.go

```diff
@@ -0,0 +1,256 @@
+		return c, true, args
+	}
+	for _, sub := range c.commands {
+		if slices.Contains(sub.Names(), args[0]) {
+			return sub.Lookup(args[1:])
+		}
+	}
+	return c, false, args
+}
+
+// The following AddFlag* methods are syntax sugar to enable flags to be added in the same line as the Command is created.
+// Not all flag types are supported yet, only the ones used in the current implementation.
+
+func (c *Command) AddFlagBool(p *bool, name string, value bool, usage string) *Command {
+	c.flags.BoolVar(p, name, value, usage)
+	return c
+}
+
+func (c *Command) AddFlagString(p *string, name string, value string, usage string) *Command {
+	c.flags.StringVar(p, name, value, usage)
+	return c
+}
+
+func (c *Command) AddFlagFunc(name string, usage string, fn func(string) error) *Command {
+	c.flags.Func(name, usage, fn)
+	return c
+}
+
+// Flags returns all flags added to this command, as well as its parent hierarchy
+func (c *Command) Flags() []*flag.Flag {
+	var flags = []*flag.Flag{}
+	c.VisitAllFlags(func(f *flag.Flag) {
+		flags = append(flags, f)
+	})
+	return flags
+}
+
+// VisitAllFlags visits all flags in the command, including those of its parent hierarchy
+func (c *Command) VisitAllFlags(fn func(f *flag.Flag)) {
+	c.flags.VisitAll(fn)
+	if c.parent != nil {
+		c.parent.VisitAllFlags(fn)
+	}
+}
+
+func (c *Command) Run(rootConfig *Config, cmdLine *CommandLine) error {
+	if !c.Runnable() {
+		return fmt.Errorf("command sidekick %s is not runnable", c.LongName())
+	}
+
+	return c.action(rootConfig, cmdLine)
+}
+
+func (c *Command) ParseCmdLine(args []string) (*CommandLine, error) {
+	if c.parent != nil {
+		c.parent.VisitAllFlags(func(f *flag.Flag) {
+			c.flags.Var(f.Value, f.Name, f.Usage)
+		})
+	}
+
+	if err := c.flags.Parse(args); err != nil {
+		return nil, err
+	}
+
+	return &CommandLine{
+		Command:             args,
+		ProjectRoot:         flagProjectRoot,
+		SpecificationFormat: format,
+		SpecificationSource: source,
+		ServiceConfig:       serviceConfig,
+		Source:              sourceOpts,
+		Language:            flagLanguage,
+		Output:              output,
+		Codec:               codecOpts,
+		DryRun:              dryrun,
+	}, nil
+}
+
+func NewCommand(
+	usageLine string,
+	shortDescription string,
+	longDescription string,
+	parent *Command,
+	action func(rootConfig *Config, cmdLine *CommandLine) error,
+) *Command {
+	if len(usageLine) == 0 {
+		panic("command usage line cannot be empty")
+	}
+	if !strings.HasPrefix(usageLine, "sidekick") {
+		panic(fmt.Sprintf("command usage line must start with sidekick, got: %s", usageLine))
+	}
+	if len(shortDescription) == 0 {
+		panic("command short description cannot be empty")
+	}
+	if parent == nil {
+		panic("command must have a parent")
+	}
+	c := newCommand(usageLine, shortDescription, parent)
+	c.longDescription = longDescription
+	c.action = action
+	return c
+}
+
+func newCommand(usageLine string, shortDescription string, parent *Command) *Command {
```

Comment:
Maybe "newRootCommand" is a better name?  Having two functions (`newCommand` and `NewCommand`) seems too subtle, and making the function private does not have much of an effect, as all the commands are in the same package.

Reply:
Yeah, I agree there is no benefit in making it private. I do think that as the app grows, it would make sense to move each command to its own package (like the Go CLI does). And in that case only `CmdSidekick` would have access to the private constructor. 

I didn't name it `newRootCommand` because I reuse the function for every command that is created, but have additional checks and logic in the exported function. I don't think it would be too bad to merge these two functions together and hardcode a check that only allows a command named "sidekick" to have no parent.

Reply:
I've merged the two constructors into one, and added a check to only allow the "sidekick" command to have no parent. I think the result is much cleaner now.